src/Utilities/FileSystem.cpp

```diff
@@ -0,0 +1,291 @@
+#include "Parallel/Printf.hpp"
+
+namespace filesystem {
+
+std::string dirname(const std::string& path) {
+  auto* file_path = new char[path.length() + 1];
+  strncpy(file_path, path.c_str(), path.length() + 1);
+  // The pointer from ::dirname is not deleted since it aliases with file_path
+  char* parent_dir_name = ::dirname(file_path);
+  std::string return_name(parent_dir_name);
+  delete[] file_path;
+  return return_name;
+}
+
+std::string get_file_name(const std::string& file_path) {
+  if (file_path.empty()) {
+    ERROR("Received an empty path");
+  }
+  if (file_path.find('/') == std::string::npos) {
+    // Handle file names such as 'dummy.txt' or '.dummy.txt'
+    return file_path;
+  }
+  std::smatch match{};
+  std::regex file_name_pattern{R"(^.*/(.+\..+))"};
+  auto regex_matched = std::regex_search(file_path, match, file_name_pattern);
+  if (not regex_matched) {
+    ERROR("Failed to find a file in the given path: '" << file_path << "'");
+  }
+  return match[1];
+}
+
+std::string get_absolute_path(const std::string& rel_path) {
+  char* name = realpath(rel_path.c_str(), nullptr);
+  if (nullptr == name) {
+    if (ENAMETOOLONG == errno) {
+      // LCOV_EXCL_START
+      ERROR(
+          "Failed to convert to absolute path because the resulting name is "
+          "too long. Relative path is: '"
+          << rel_path << "'.");
+      // LCOV_EXCL_STOP
+    } else if (EACCES == errno) {
+      // LCOV_EXCL_START
+      ERROR(
+          "Failed to convert to absolute path because one of the components of "
+          "the path does not have proper read access. Relative path is: '"
+          << rel_path << "'.");
+      // LCOV_EXCL_STOP
+    } else if (ENOENT == errno) {
+      ERROR(
+          "Failed to convert to absolute path because one of the path "
+          "components does not exist. Relative path is: '"
+          << rel_path << "'.");
+      // LCOV_EXCL_START
+    } else if (ELOOP == errno) {
+      ERROR(
+          "Failed to convert to absolute path because the maximum number of "
+          "symlinks was in the path. Relative path is: '"
+          << rel_path << "'.");
+      // LCOV_EXCL_STOP
+    }
+    ERROR("Failed to get absolute path for an unknown reason.");
+  }
+  std::string return_name(name);
+  delete[] name;
+  return return_name;
+}
+
+bool create_directory(std::string dir, const double wait_time,
+                      const size_t num_tries, const bool terminate_on_failure) {
+  // each time we fail to create a directory, we increase the wait_time by this
+  // factor
+  static constexpr double wait_time_increase_factor = 1.1;
+  if (dir.empty()) {
+    ERROR("Cannot create a directory that has no name");
+  }
+  // Remove the trailing slashes on the directory name
+  while (not dir.empty() and dir[dir.size() - 1] == '/') {
+    dir.erase(dir.size() - 1);
+  }
+  if (dir.empty()) {
+    ERROR("Cannot create the root directory '/' (possibly multiple slashes)");
```

Comment:
This should be a success case according to the documentation, since / presumably exists.